### ISSUE TITLE
Save on last player exit might be misbehaving

### ISSUE THREAD

**Issue by hakusaro** [Pryaxis/TShock#905]
https://tshock.co/xf/index.php?threads/quick-questions-thread-ask-your-how-do-i-and-where-is-questions-here.860/page-43#post-39413

Should be tested with varying amounts of players disconnecting until 1 is remaining and seeing if it saves in a loop.

**Comment by WhiteXZ** [comment]
Afaik there's nothing in the codebase that should cause this, unless the SaveManager is doing weird shit that it's never done before.

**Comment by WhiteXZ** [comment]
Could not reproduce with stock server

**Comment by nicatronTg** [comment]
I couldn't reproduce this either.

**Comment by WhiteXZ** [comment]
Just saw this happen on a server. Still not sure how to reproduce

**Comment by bippity** [comment]
@WhiteXZ were you referring to the stuff that happened at 2015-05-22 02:13:17 in this [log](https://drive.google.com/file/d/0B4AQyecqIdqiZVdmdWxfQmRrUW8/view?usp=sharing)?

**Comment by nicatronTg** [comment]
Is this still an issue?

**Comment by chainsol** [comment]
Unable to consistently reproduce, but I'm having this happen. I'm the only player on the server, but I'm using a service that logs in to check if the server is online, and every time it does that it causes the server to save repeatedly, due to it logging in and then logging out.

**Comment by tlworks** [comment]
Additionally simply turn this feature OFF in the config to prevent it from happening, the Server auto-saves anyway.

**Comment by chainsol** [comment]
I like your moxy, but it still does the same thing. Reloaded server after turning off auto-save and I still get doubled/tripled saves. :D

**Comment by bartico6** [comment]
Bumping, still an issue.
![Proof](https://cdn.discordapp.com/attachments/304040036977082368/349168875411079168/unknown.png)

**Comment by DeathCradle** [comment]
I'm able to reproduce this with one player online, while spamming invalid connections causing [OnLeave](https://github.com/Pryaxis/TShock/blob/623815b9f68a72401af6a9f3f47ee9b893987fcb/TShockAPI/TShock.cs#L1451) to fire each time. You can do this from a web browser.

Setting `SaveWorldOnLastPlayerExit` to `false` fixes my scenario. 

![image](https://user-images.githubusercontent.com/776327/32226747-299fe30c-be95-11e7-8488-e2bfb0a12480.png)

**Comment by chainsol** [comment]
This has been a problem for a while - has anyone narrowed down the cause? I notice one of the most recent issues, #1490, might be related?

**Comment by bartico6** [comment]
I have "SaveWorldOnLastPlayerExit" to false and the above still happens.

**Comment by BurningUnder** [comment]
I have met the same question like "When I have join in the server and my friend wanted to join in ,the server will immediately Saving World data, and he will stuck on 'Find Server' until the saving have completed."
Every time we have to wait till the saving completed just like what had happened above that @DeathCradle 's picture have shown.

**Comment by hakusaro** [comment]
It would not really surprise me if this is caused by logic in Terraria saving the server (as opposed to us telling the server to save).

**Comment by DankRank** [comment]
This happens to me a lot. I managed to pinpoint the conditions for it to occur:
- `SaveWorldOnLastPlayerExit` must be true
- There must be exactly 1 player on the server
- Someone connects to the server using a wrong protocol (I use the same port for both terraria and bittorrent, so I get lots of random connections to tshock like that)

The code responsible for saving on last player exit [runs only when there's one active player left, and OnLeave is invoked](https://github.com/Pryaxis/TShock/blob/v4.3.25/TShockAPI/TShock.cs#L1412). It looked like OnLeave somehow gets called when an invalid client gets disconnected, however it's [only called when client is active](https://github.com/Pryaxis/TerrariaAPI-Server/blob/c31b0c3ca12eb638601f140f66c38a956adc1c92/TerrariaServerAPI/TerrariaApi.Server/Hooking/ServerHooks.cs#L47). I think this is most likely a terraria bug.

**Comment by QuiCM** [comment]
Thanks for the investigation. Definitely seems like a Terraria bug.

Off the top of my head, I wonder if this could be solved by something as simple as (pseudo code):
```
if (playerCount is 1 and leavingId is players[0].id)
    DoWorldSave()
```

**Comment by DankRank** [comment]
I did some more digging and it looks like `GetActivePlayerCount` checks for `Player` entities, while the TSAPI hook checks for `RemoteClient`. So, this looks like a TShock/TSAPI bug after all.

---
It works very consistentently on v4.3.25. I used netcat to simulate the invalid connections:
```sh
nc -w0 127.0.0.1 7777 >/dev/null
```

However, I couldn't reproduce it on the latest devel commit (035e9d114b4bd31d5ace1be98277a981a6fb1202, travis build 2213)
